### DISCUSSION TITLE
how to convert a rgbaImage to opencv's Mat?

### DISCUSSION THREAD

**Discussion by ProgramAPE** [image-rs/image#2307] [Q&A]
I want to use the image to take screencapture and then send the pic to opencv to analysis.
and the opencv use Mat data type. so i want to turn the image to Mat.
Did anyone can help me? ths!

**Comment by ProgramAPE** [comment]
after some research i find a way to do it . 
But i am not sure  if this is the best way to do it.
unsafe {
		src = Mat::new_rows_cols_with_data_unsafe_def(height as i32, width as i32, core::CV_8UC3, buffer.as_mut_ptr() as *mut _, )?;
}

**Reply by 197g** [reply]
This is a good way to reference external data in `cv`. However, care must be taken:

> Matrix constructors that take data and step parameters do not allocate matrix data. Instead, they just initialize the matrix header that points to the specified data, which means that no data is copied. This operation is very efficient and can be used to process external data using OpenCV functions. The external data is not automatically deallocated, so you should take care of it. 

This means the matrix must not be used after the `image`'s `buffer` is deallocated. You can call `cv::Mat::clone` to take an owning copy of the created matrix that lives independently from the `buffer`, and so does not require it to stay alive.

**Comment by konstantinzolotarev** [comment]
First of all, big thank you @ProgramAPE and @HeroicKatora, spent almost an hour on trying to convert image into `Mat` unless found this discussion.
Just in case if somebody will search for a working code in future: 

Note: I'm using `RGBa` with alpha layer. If you wouldn't need it, you will have to switch `core::CV_8UC4` to correct type.
```rust
// Converts rawImage into opencv::Mat
fn into_mat(image: &RgbaImage) -> Mat {
    unsafe {
        Mat::new_rows_cols_with_data_unsafe_def(
            image.height() as i32,
            image.width() as i32,
            core::CV_8UC4,
            image.as_raw().clone().as_mut_ptr() as *mut _,
        )
        .unwrap()
    }
}
```
